Quotient/room.cpp

```diff
@@ -2844,6 +2848,40 @@ Room::Changes Room::Private::addNewMessageEvents(RoomEvents&& events)
             if (const auto* evt = it->viewAs<CallEvent>())
                 emit q->callEvent(q, evt);
 
+    for (auto it = from; it != syncEdge(); ++it) {
+        //TODO is the order here reversed?
+        if (it->event()->senderId() == connection->userId()) {
+            if (const auto* evt = it->viewAs<RoomMessageEvent>()) {
+                if (evt->rawMsgtype() == "m.key.verification.request"_ls) {
+                    keyVerificationSessions[evt->id()] = pendingKeyVerificationSession;
+                    pendingKeyVerificationSession->setRequestEventId(evt->id());
+                    pendingKeyVerificationSession = nullptr;
+                }
+                continue;
+            } else {
+                continue;
+            }
+        }
+        if (const auto* evt = it->viewAs<RoomMessageEvent>()) {
+            if (evt->rawMsgtype() == "m.key.verification.request"_ls) {
+                auto session = new KeyVerificationSession(evt->senderId(), connection, q->usesEncryption(), {}, evt->contentPart<QStringList>("methods"_ls), evt->originTimestamp(), evt->contentPart<QString>("from_device"_ls), q, evt->id());
+                emit connection->newKeyVerificationSession(session);
+                //TODO connect delete from sessions when finished
+                keyVerificationSessions[evt->id()] = session;
+            }
+        }
+        if (it->event()->matrixType().startsWith("m.key.verification."_ls)) {
+            auto event = loadEvent<KeyVerificationEvent>(it->event()->fullJson());
```

Comment:
This really should not be so complicated.
```suggestion
        if (auto event = it->viewAs<KeyVerificationEvent>()) {
```
But make sure to derive `KeyVerificationEvent` from `RoomEvent` so that normal room event deserialisation could use these classes.